src/shared/convert.js

```diff
@@ -1,35 +1,49 @@
 // @flow
 
-import type {ColorMatchType} from "./colors";
+import type {ColorMatchType} from './colors';
 import {colorArray, formats} from './colors';
 import type {TinyColor} from 'tinycolor2';
 import tinycolor from 'tinycolor2';
-import {ObjectMap} from "./generic";
+import {ObjectMap} from './generic';
+import diff from 'color-diff';
 
-function diffColors(from: string, to: string): number {
-  const materialRgb = tinycolor(from).toRgb();
-  const colorRgb = tinycolor(to).toRgb();
-  const diffColor = (key) => Math.abs(materialRgb[key] - colorRgb[key]);
-  return diffColor('r') + diffColor('g') + diffColor('b');
-}
+// `color-diff` only allows uppercase {R,G,B}
+// TODO: remove upcasing after submitting a PR for downcase
+type UpcaseRgbType = {
+  R: number,
+  G: number,
+  B: number
+};
+const hexToUpcaseRgb = function (hex: string): UpcaseRgbType {
+  const rgbColor = tinycolor(hex).toRgb();
+  return {R: rgbColor.r, G: rgbColor.g, B: rgbColor.b};
+};
 
-export function convert(color: string): ColorMatchType {
-  const colorMatch = colorArray.reduce((match: { diffVal: number, color: ColorMatchType }, colorItem: ColorMatchType) => {
-    const diffVal = diffColors(color, colorItem.value);
-    if (diffVal < match.diffVal) {
-      return {
-        diffVal,
-        color: colorItem
-      }
-    } else {
-      return match;
-    }
-  }, {
-    diffVal: Infinity,
-    color: ({}: any)
-  });
-  return colorMatch.color;
-}
+// Hoist upcase RGB objects so they're only created once (on first use)
+let colorMapRgb: Map<UpcaseRgbType, ColorMatchType>;
+let palette: Array<UpcaseRgbType>;
+
+export function convert(fromHex: string): ColorMatchType {
+  // Convert input color to upcase RGB object
+  const fromRgb = hexToUpcaseRgb(fromHex);
+
+  if (!colorMapRgb) {
+    // Create `Map` with pairs of `upcaseRgb: paletteColorObject`
+    colorMapRgb = new Map(
+      colorArray.map((colorObject) => [hexToUpcaseRgb(colorObject.value), colorObject])
+    );
```

Comment:
Let's add a comment that explains that `hexToUpcase` is not a string

Reply:
I modified a couple of comments to specifically mention their type while mentioning that they're objects. 😄